gb-roms/src/controllers/mbc2.rs

```diff
@@ -0,0 +1,170 @@
+            let address = address - 0x4000;
+            Ok(rom[address])
+        }
+    }
+
+    fn get_selected_rom(&self, is_root_bank: bool) -> &[u8; MBC2::ROM_SIZE] {
+        let index: usize = if is_root_bank {
+            0
+        } else {
+            self.regs.rom_number.into()
+        };
+
+        &self.rom_bank[index]
+    }
+
+    fn write_ram(&mut self, v: u8, addr: Box<dyn Address<Area>>) -> Result<(), Error> {
+        if !self.regs.ram_enabled {
+            return Err(Error::new_segfault(addr));
+        }
+        let address = addr.get_address();
+        self.ram_bank[address] = v & 0xF;
+        Ok(())
+    }
+
+    fn read_ram(&self, addr: Box<dyn Address<Area>>) -> Result<u8, Error> {
+        if !self.regs.ram_enabled {
+            return Err(Error::new_segfault(addr));
+        }
+        let address = addr.get_address();
+        Ok(self.ram_bank[address])
+    }
+}
+
+impl FileOperation<Area> for MBC2 {
+    fn write(&mut self, v: u8, addr: Box<dyn Address<Area>>) -> Result<(), Error> {
+        match addr.area_type() {
+            Area::Rom => self.write_rom(v, addr),
+            Area::Ram => self.write_ram(v, addr),
+            _ => Err(Error::new_bus_error(addr)),
+        }
+    }
+
+    fn read(&self, addr: Box<dyn Address<Area>>) -> Result<u8, Error> {
+        match addr.area_type() {
+            Area::Rom => self.read_rom(addr),
+            Area::Ram => self.read_ram(addr),
+            _ => Err(Error::new_bus_error(addr)),
+        }
+    }
+}
+
+struct MBC2Reg {
+    /// Enable READ/WRITE operation on RAM
+    ram_enabled: bool,
+    /// Select ROM bank id in area 0x4000-0xbfff
+    rom_number: u8,
+}
+
+impl Default for MBC2Reg {
+    fn default() -> Self {
+        Self {
+            ram_enabled: false,
+            rom_number: 1,
+        }
+    }
+}
+#[derive(serde::Serialize, serde::Deserialize)]
+struct Mbc2Data {
+    ram_bank: Vec<u8>,
+}
+
+impl From<[u8; MBC2::RAM_SIZE]> for Mbc2Data {
+    fn from(bank: [u8; MBC2::RAM_SIZE]) -> Self {
+        Self {
+            ram_bank: bank.iter().map(|byte| *byte).collect(),
+        }
+    }
+}
+
+impl Controller for MBC2 {
+    fn save<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let data = Mbc2Data::from(self.ram_bank.clone());
+        data.serialize(serializer)
+    }
+
+    fn load<'de, D>(&mut self, deserializer: D) -> Result<(), D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+        use std::convert::TryFrom;
+
+        let data = Mbc2Data::deserialize(deserializer)?;
+
+        self.ram_bank = <[u8; MBC2::RAM_SIZE]>::try_from(data.ram_bank).map_err(|faulty| {
+            Error::invalid_length(
+                faulty.len(),
+                &format!("a ram bank size of size {}", MBC2::RAM_SIZE).as_str(),
+            )
+        })?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test_mbc2 {
+    //TODO: implement test
+}
```

Comment:
to remove if let empty

Reply:
See commit: https://github.com/Ragarnoy/GBMU/pull/127/commits/203feb68bd90a8434cab018bb8e7634fb46f9e52